### ISSUE TITLE
[Help] 如何进行人类反馈强化学习微调?

### ISSUE THREAD

**Issue by raymond00000** [THUDM/ChatGLM-6B#1337]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

你好，我有一些对话数据，这些数据是打了分数的，我想问一下如何用我手上的数据进行“人类反馈强化学习”微调？我会尝试 P-Tuning v2，但更想尝试“人类反馈强化学习”的微调。谢谢！

### Steps To Reproduce

NA

### Environment

```markdown
OS: Ubuntu 20.04
    Python: 3.8
    Transformers: 4.26.1
    PyTorch: 1.12
    CUDA Support: True
```

**Comment by tomcat123a** [comment]
https://github.com/shibing624/MedicalGPT
参考这个项目，预训练，指令微调，rm模型训练，ppo都有现成的。

**Comment by raymond00000** [comment]
谢谢你的回馈，我见这个medicalGPT很有参考价值。 
但 medicalGPT 的 read me 显示 multi-round dialogue data fine-tuning method是 to-do. 
medicalGPT 好像是 Q and A 的 GPT。
我是想把ChatGLM-6B 微调后还是 chat model。

medicalGPT中， RL（强化学习）数据集格式如下：
json文件，每行一个样本，每个样本包含以下字段：
{"instruction": "text1", "input": "text2", "output": "text3"}
好像不支持 多句 dialogue 的数据。

https://github.com/shibing624/MedicalGPT/wiki/%E8%AE%AD%E7%BB%83%E7%BB%86%E8%8A%82%E8%AF%B4%E6%98%8E

**Comment by tomcat123a** [comment]
多轮很容易，前面加“问：xxxx\n 答：xxxx\n 问：xxxx\n 答：xxxx\n

**Comment by tomcat123a** [comment]
参考它聊天输出的代码。

**Comment by raymond00000** [comment]
是这样子吗？

{“instruction”： “你是一个秘书.. 老板今天的事项..”， “input”： “老板问：今天有多少个会议”， “output”： “秘书答：你今天早上要3个开会”}
{“instruction”： “你是一个秘书.. 老板今天的事项..”， “input”： “老板问：今天有多少个会议\n 秘书答：你今天早上要3个开会\n 老板问：第一个会是什么时候”， “output”： “11时”}

还是分开？

{“instruction”： “你是一个秘书.. 老板今天的事项..”， “input”： “老板问：今天有多少个会议”， “output”： “秘书答：你今天早上要3个开会”}
{“instruction”： “你是一个秘书.. 老板今天的事项..”， “input”： “老板问：第一个会是什么时候”， “output”： “11时”}

谢谢你的见解!

**Comment by tomcat123a** [comment]
不是。麻烦看下它history传入之后的代码，知道多轮对话如何inference的，就知道如何训练了。你发的是单指令微调的，是一轮的，跟多轮关系不大。

**Comment by tomcat123a** [comment]
git首页就有。

**Comment by valkryhx** [comment]
> 多轮很容易，前面加“问：xxxx\n 答：xxxx\n 问：xxxx\n 答：xxxx\n

是不是sample仍然是instruction
input output 三个字段 只不过instruction 是历史对话history通过一定的模板[chatglm源代码中写了这个模板]拼起来 并且拼接上最后一个提问句 output字段写上答案 基本上应该是这个形式了吧？

**Comment by valkryhx** [comment]
其实还是把多轮对话转化成了单轮对话，没什么技巧。只不过多轮对话包含历史信息而已，不知道我理解的对不对？@tomcat123a 另外想请问一下，你说的那个给的首页是哪个项目给的首页？我在shibing64这个项目上确实没有看到多轮对话的例子，能否麻烦你指出一下具体位置

**Comment by tomcat123a** [comment]
具体位置在ptuning里面。https://github.com/THUDM/ChatGLM-6B/tree/075387c735c4d96ab735d6087c96b1087410cccf/ptuning#对话数据集

**Comment by valkryhx** [comment]
> 具体位置在ptuning里面。[https://github.com/THUDM/ChatGLM-6B/tree/075387c735c4d96ab735d6087c96b1087410cccf/ptuning#对话数据集](https://github.com/THUDM/ChatGLM-6B/tree/075387c735c4d96ab735d6087c96b1087410cccf/ptuning#%E5%AF%B9%E8%AF%9D%E6%95%B0%E6%8D%AE%E9%9B%86)

可以参考这段代码 其实就是把history按照模板拼起来 再和最后一轮的query再拼接作为最终的提问prompt
